### PR TITLE
Implement kickout for chunk producers

### PR DESCRIPTION
**Pull request by bowenwang1996**
Validators will now be kicked out for not producing sufficient number of chunks.

### CODE REVIEW

File: chain/epoch_manager/src/lib.rs [hunk 1]
```diff
@@ -69,14 +69,87 @@ impl EpochManager {
...
+            {
+                validator_kickout.insert(account_id.clone());
+                continue;
+            }
+
+            // Given the number of blocks we plan to have in one epoch, the following code should not overflow
+            validator_online_ratio.insert(
+                account_id.clone(),
+                (
+                    num_blocks * total_chunks_expected + expected_blocks * total_chunks_produced,
```
[the remark this answers is not in the document]
Comment (bowenwang1996) [reply chain/epoch_manager/src/lib.rs]: Average between online ratio for blocks and chunks. Do we want something else?

File: core/primitives/src/block.rs [hunk 2]
```diff
@@ -390,6 +402,7 @@ impl Block {
                 approval_sigs,
                 total_weight,
                 validator_proposals,
+                chunk_mask,
```
[the remark this answers is not in the document]
Comment (bowenwang1996) [reply core/primitives/src/block.rs]: I wanted to cache it but yeah now I think it's a bad idea

File: near/src/runtime.rs [hunk 3]
```diff
@@ -105,31 +105,36 @@ impl NightshadeRuntime {
...
+                    set_account(state_update, &account_id, &account);
+                }
             }
         }
-        if let Some(mut protocol_treasury_account) =
-            get_account(state_update, &self.genesis_config.protocol_treasury_account)
-        {
+        if self.account_id_to_shard_id(&self.genesis_config.protocol_treasury_account) == shard_id {
+            let mut protocol_treasury_account =
+                get_account(state_update, &self.genesis_config.protocol_treasury_account).unwrap();
```
[the remark this answers is not in the document]
Comment (bowenwang1996) [reply near/src/runtime.rs]: Okay but protocol treasury cannot be deleted right?

File: near/src/runtime.rs [hunk 4]
```diff
@@ -105,31 +105,36 @@ impl NightshadeRuntime {
...
-                    "account {} stake {} max_of_stakes: {}",
-                    account_id, account.stake, max_of_stakes
-                );
-                assert!(account.stake >= max_of_stakes, "FATAL: staking invariance does not hold");
-                let return_stake = account.stake - max_of_stakes;
-                account.stake -= return_stake;
-                account.amount += return_stake;
+                    assert!(
+                        account.stake >= max_of_stakes,
+                        "FATAL: staking invariant does not hold"
```
[the remark this answers is not in the document]
Comment (bowenwang1996) [reply near/src/runtime.rs]: `account.stake - max_of_stakes` is the expected return stake.

File: near/src/runtime.rs [hunk 5]
```diff
@@ -430,7 +436,9 @@ impl RuntimeAdapter for NightshadeRuntime {
                 &mut state_update,
             )
             .map_err(|e| Error::from(ErrorKind::ValidatorError(e.to_string())))?;
+            state_update.commit();
```
[the remark this answers is not in the document]
Comment (bowenwang1996) [reply near/src/runtime.rs]: It was a mistake.

File: chain/epoch_manager/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ilblackdragon) [reply chain/epoch_manager/src/lib.rs]: Ok for now. Can discuss a bit in person next week.

File: near/src/runtime.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ilblackdragon) [reply near/src/runtime.rs]: Protocol treasury - no

File: near/src/runtime.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ilblackdragon) [reply near/src/runtime.rs]: I mean change the error message to actually show what was the erroring values

**Pull request merged**